File: 1/1000s/12s/1298/MaxCandiesYouCanGetFromBoxes.java
/*
https://leetcode.com/problems/maximum-candies-you-can-get-from-boxes/description/?envType=daily-question&envId=2025-06-03
*/

class Solution {
    public int maxCandies(int[] status, int[] candies, int[][] keys, int[][] containedBoxes, int[] initialBoxes) {
        int lengthOfStatus=status.length, resultantCandies=0;
        boolean[] availableBoxes=new boolean[lengthOfStatus];
        boolean[] canOpenBox=new boolean[lengthOfStatus];
        boolean[] seenBoxes=new boolean[lengthOfStatus];
        LinkedList<Integer>llOfBoxes=new LinkedList<>();
        for(int i=0; i<lengthOfStatus; ++i){
            availableBoxes[i]=false;
            canOpenBox[i]=false;
            seenBoxes[i]=false;
            if(status[i]==1){
                canOpenBox[i]=true;
            }
        }
        for(int iniBoxes: initialBoxes){
            availableBoxes[iniBoxes]=true;
            if(canOpenBox[iniBoxes]){
                resultantCandies+=candies[iniBoxes];
                seenBoxes[iniBoxes]=true;
                llOfBoxes.add(iniBoxes);
            }
        }
        while(llOfBoxes.size()>=1){
            int currentLLSize=llOfBoxes.size();
            while(currentLLSize-- >0){
                int currentBox=llOfBoxes.pollFirst();
                for(int currKey: keys[currentBox]){
                    canOpenBox[currKey]=true;
                    if(!seenBoxes[currKey]&&availableBoxes[currKey]){
                        resultantCandies+=candies[currKey];
                        seenBoxes[currKey]=true;
                        llOfBoxes.add(currKey);
                    }
                }
                for(int curContBox: containedBoxes[currentBox]){
                    availableBoxes[curContBox]=true;
                    if(!seenBoxes[curContBox]&&canOpenBox[curContBox]){
                        resultantCandies+=candies[curContBox];
                        seenBoxes[curContBox]=true;
                        llOfBoxes.add(curContBox);
                    }
                }
            }
        }
        return resultantCandies;
    }
}

File: 1/1000s/12s/1298/MaxCandiesYouCanGetFromBoxes.cpp
/*
https://leetcode.com/problems/maximum-candies-you-can-get-from-boxes/description/?envType=daily-question&envId=2025-06-03
*/

class Solution {
public:
    int maxCandies(vector<int>& status, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes, vector<int>& initialBoxes) {
        int sz=status.size(), res=0;
        vector<bool>hasBoxes(sz, false), canOpen(sz, false), seenBoxes(sz, false);
        queue<int>q;
        for(int i=0; i<sz; ++i){
            if(status[i]==1){
                canOpen[i]=true;
            }
        }
        for(auto &iniBoxes: initialBoxes){
            hasBoxes[iniBoxes]=true;
            if(canOpen[iniBoxes]){
                res+=candies[iniBoxes];
                seenBoxes[iniBoxes]=true;
                q.push(iniBoxes);
            }
        }
        while(!q.empty()){
            int qsz=q.size();
            while(qsz--){
                int curBox=q.front();
                q.pop();
                for(int curKey: keys[curBox]){
                    canOpen[curKey]=true;
                    if(!seenBoxes[curKey]&&hasBoxes[curKey]){
                        res+=candies[curKey];
                        seenBoxes[curKey]=true;
                        q.push(curKey);
                    }
                }
                for(int curContBox: containedBoxes[curBox]){
                    hasBoxes[curContBox]=true;
                    if(!seenBoxes[curContBox]&&canOpen[curContBox]){
                        res+=candies[curContBox];
                        seenBoxes[curContBox]=true;
                        q.push(curContBox);
                    }
                }
            }
        }
        return res;
    }
};